lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +48,141 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(end_of_word, str):
+        return None
+    if not isinstance(start_of_word, str) and start_of_word is not None:
+        return None
+
+    dict_freq = {}
+    words_list = text.split()
+    for word in words_list:
+        prepared_word = prepare_word(word, start_of_word, end_of_word)
+        if prepared_word is None:
+            return None
+        if prepared_word not in dict_freq:
+            dict_freq[prepared_word] = words_list.count(word)
+
+    return dict_freq
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_dict = {}
+
+    for word in word_frequencies:
+        for index, token in enumerate(word):
+            if index == len(word) - 1:
+                break
+            next_token_index = index + 1
+            pair = (token, word[next_token_index])
+            pair_dict[pair] = 0
```

Comment:
Consider accumulating pairs frequency in this cycle, you don't need the second cycle for that

Reply:
fixed